### ISSUE TITLE
Top-Down OOP increases coupling?

### ISSUE THREAD

**Issue by thatseant** [nus-cs2113-AY2021S1/forum#91]
It has been frequently suggested during the course that output be extracted out to Ui and messages class and parsing functionalities be extracted to Parser class. 

Doesn't this approach mean for every command added, we need to change Ui, Parser, Command, Messages and object classes extensively? Does this indicate high coupling? As most of our development are adding commands over time, are we changing too many classes every time we do that?

Would a better approach be to put most of the logic for every command in Command classes? Then, adding a command would mostly require changes only to that specific command class. Other teammates can work on other commands with lesser conflicts.

**Comment by kstonekuan** [comment]
I think this is a very valid concern and my team has also faced this issue with many merge conflicts over the Ui class. However, I think putting a lot of logic into the Command classes would start to violate SRP and as a result the code becomes more procedural in nature rather that being object-oriented.

**Comment by hughjazzman** [comment]
Not necessarily. If you have proper abstraction within the `Parser` class, a new command could simply use an existing method to prepare the arguments. If your `Parser`, `Command`, `Messages` don't need to be aware of `Ui` as discussed in #88, then a new command would just be a new `Command` subclass, with `Messages` being associated with the `Command`.

**Comment by okkhoy** [comment]
Valid concern and reasonably good arguments and suggestions on how to go about it. 
Do take a look at AB-3 codebase to get an idea.
